File: src/main/java/leetcode/YY20/MM09/DD13_InsertInterval.java
package leetcode.YY20.MM09;

import java.util.ArrayList;
import java.util.List;

public class DD13_InsertInterval {
    public static void main(String[] args) {

    }

    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> result = new ArrayList();

        for (int[] interval : intervals) {
            if (interval[1] < newInterval[0]) {
                result.add(interval);
            } else if (newInterval[1] < interval[0]) {
                result.add(newInterval);
                newInterval = interval;
            } else {
                newInterval[0] = Math.min(newInterval[0], interval[0]);
                newInterval[1] = Math.max(newInterval[1], interval[1]);
            }
        }

        result.add(newInterval);

        return result.toArray(new int[result.size()][]);
    }
}

File: src/main/kotlin/leetcode/YY20/MM09/DD13_InsertInterval.kt
import kotlin.math.max
import kotlin.math.min

fun main() {
    for (a in insert(arrayOf(intArrayOf(1, 3), intArrayOf(6, 9)), intArrayOf(2, 5))) {
        println(a.contentToString())
    }
}

fun insert(intervals: Array<IntArray>, newInterval: IntArray): Array<IntArray> {
    val result = mutableListOf<IntArray>()
    var newInterval = newInterval

    for (interval in intervals) {
        if (interval[1] < newInterval[0]) {
            result.add(interval)
        } else if (newInterval[0] < interval[0]) {
            result.add(newInterval)
            newInterval = interval
        } else {
            newInterval[0] = min(newInterval[0], interval[0])
            newInterval[1] = max(newInterval[1], interval[1])
        }
    }

    result.add(newInterval)

    return result.toTypedArray()
}